db/crud.go

```diff
@@ -1882,6 +1882,9 @@ func (col *DatabaseCollectionWithUser) documentUpdateFunc(ctx context.Context, d
 
 	unusedSequences, err = col.assignSequence(ctx, previousDocSequenceIn, doc, unusedSequences)
 	if err != nil {
+		if errors.Is(err, base.ErrMaxSequenceReleasedExceeded) {
+			base.WarnfCtx(ctx, "Doc %s / %s had an existing sequence %d that is more than %d larger than expected. Document update will be cancelled. This may indicate documents being migrated between databases by an external process.", base.UD(newDoc.ID), prevCurrentRev, doc.Sequence, MaxSequencesToRelease)
```

Comment:
Since this indicates manual intervention to strip doc metadata, I think this is one of few times that Error level is appropriate to log at.

Also reworded log message and used `previousDocSequenceIn` to be safe (in case `col.assignSequence` changed `doc.Sequence`

```suggestion
			base.ErrorfCtx(ctx, "Doc %s / %s had a much larger sequence (%d) than the current sequence number. Document update will be cancelled, since we don't want to allocate sequences to fill a gap this large. This may indicate document metadata being migrated between databases where it should've been stripped and re-imported.", base.UD(newDoc.ID), prevCurrentRev, previousDocSequenceIn, MaxSequencesToRelease)
```

Reply:
I agree, moved it to error. I've left it as doc.Sequence as upon using `previousDocSequenceIn` I was getting 0 sequence values in the log line. It looks like from inside `updateAndReturnDoc` this value is always 0 when its passed into `documentUpdateFunc`. It seems to only be used post `documentUpdateFunc` after a sequence is allocated. When i fixed this (to the actual current doc seq value) I had a few test failures, looking at the code inside `assignSequence` it drops into this https://github.com/couchbase/sync_gateway/blob/5bd0d640ad377b80a2c385023d0381d10b0919a3/db/crud.go#L1669 when the sequence that has been passed in is non zero. The more I look at this code I feel like something is not right here... but i don't want to change this code too much given we hope to get 3.2.2.1 out with minimal changes and I though we want to keep risk low? Happy to hear your thoughts though.

Anyway I think its fine as `doc.Sequence` as we won't have assigned the seq to the doc if we hit this at that point.

Reply:
Sounds good, agree with minimizing the changes

Reply:
Upon review of the code `previousDocSequenceIn` being 0 when passed into the this function makes sense. Given the callback this function sits in code that can be invoked multiple times (for example, if we got CAS failure that we should retry on). It's used for `unused_sequence` handling so the change cache doesn't get hung up on not receiving sequences that were allocated but are unusable.

May file an enhancement for future to improve variable naming/add some comments around the doc sequence variable being passed down from the callback to ensure no confusion in future.